### PR TITLE
Python: postgres memory store

### LINKED ISSUE #1270: Python: Add Postgres with pgvector as a Memory Store for Python
I would like to add Postgres with pgvector as a Memory Store to the Python implementation to give parity for the addition to .net in #945

### PR DESCRIPTION
**Pull request by cschadewitz**
### Description
Implemention of the MemoryStore to use a Postgresql (with pgvector extension) to resolve #1270 

Taking a different approach from the .net implementation:
Collections as postgres schemas as opposed to collection being another column in the table, willing to change to match but would love to have a discussion comparing the options. 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by cschadewitz** [comment]
@microsoft-github-policy-service agree

**Comment by cschadewitz** [comment]
Working on change to conform to new Memory Record field

File: python/tests/integration/connectors/memory/test_postgres.py
Comment (dmytrostruk) [inline python/tests/integration/connectors/memory/test_postgres.py:110]: Remove unused code:
```suggestion
```

File: python/tests/integration/connectors/memory/test_postgres.py
Comment (dmytrostruk) [inline python/tests/integration/connectors/memory/test_postgres.py:112]: I would not rely on amount of collections in test database, but verify if `test_collection` is in `result`:
```suggestion
    assert "test_collection" in result
```

File: python/tests/integration/connectors/memory/test_postgres.py
Comment (dmytrostruk) [inline python/tests/integration/connectors/memory/test_postgres.py:149]: Should we also verify how embedding vector was stored by checking actual vector values?

File: python/tests/integration/connectors/memory/test_postgres.py
Comment (dmytrostruk) [inline python/tests/integration/connectors/memory/test_postgres.py:184]: Same as for collections, I would add extra `get` operation just to double-check that deletion actually happened:
```suggestion
    await memory.create_collection_async("test_collection", 2)
    await memory.upsert_async("test_collection", memory_record1)
    
    result = await memory.get_async(
        "test_collection", memory_record1._id, with_embedding=True
    )
    assert result is not None
    
    await memory.remove_async("test_collection", memory_record1._id)
    with pytest.raises(KeyError):
        _ = await memory.get_async(
            "test_collection", memory_record1._id, with_embedding=True
        )
```

**Review by dmytrostruk — COMMENTED** [review]
@cschadewitz Thank you for contribution! I left initial comments. As soon as they are resolved, we will test it locally and if everything works as expected - merge it to main! 😃

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:24]: Makes sense, I had originally set this up to allow for delineation of the collections to be at the schema level. Since the direction is to go with table level delineation, I will make the necessary changes accordingly.

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:64]: Will implement full parameterization

File: python/semantic_kernel/connectors/memory/pinecone/pinecone_memory_store.py
Comment (cschadewitz) [inline python/semantic_kernel/connectors/memory/pinecone/pinecone_memory_store.py:1]: fixed param ordering in doc string

**Comment by cschadewitz** [comment]
@dmytrostruk All PR comments so far should be resolved, I have also fixed one issue in the doc string from the pinecone mem store and have added the dependencies to the .toml file. 

Let me know if anything else comes up.

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (JadynWong) [inline python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:94]: The C# implementation does not execute this statement anymore, because it only needs to be executed once after the database is created, and most importantly it seems that not all hosting can execute this statement, for example [Azure Cosmos DB for PostgreSQL](https://learn.microsoft.com/en-us/azure/cosmos-db/postgresql/howto-use-pgvector).

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (JadynWong) [inline python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:103]: C# uses BIGINT type to unix to store timestamp, I am considering whether to use TIMESTAMP to store it. The postgres timestamp requires that it be stored in UTC time, and when you query it again, it returns the utc time, with no way of knowing the actual time zone when it was stored. How do you think about it?

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:94]: Interesting, I didnt know that there were some hosting options with different statements for this. I can certainly remove it. Thanks for pointing this out! I have seen some of the work you have done/are doing against the C# side for the Postgres memory store. I'll have to take a look at what you currently have in the pipeline. Thanks @JadynWong

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:103]: This was something that I saw when I was looking at the most recent version of the C# implementation, so I am really glad that you brought this up! You are correct based on the docs, Postgres has the concept of both timestamp without time zone and timestamp with time zone, and this does not affect the storage as both are ultimately stored as 8-byte integers in UTC. 

[Time Stamps](https://www.postgresql.org/docs/current/datatype-datetime.html#:~:text=8.5.1.3.%C2%A0Time%20Stamps)

This would mean that we would need to be contextually aware of the timezone of the client (which sounds like we are in the C# implementation) and sprinkle in the `timezone(zone, timestamp)` or 'AT TIME ZONE' conversion functions to push the conversion work onto the db.

[Time Stamp Functions](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT)

Ultimately, I think that we should go with `timestamp with time zone`and use the built-in functions where appropriate. Do you agree?

Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py]: Initial stab at handling this is in the conde now.

Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py]: I have changed the `MemoryRecord `implementation to use `datetime` instead of `str` for the `timestamp`. With this in place I have changed the `PostresqlMemoryStore` and tests to account for this. 

Additionally, I have made the necessary changes to the `ChromaMemoryStore` and `PineconeMemoryStore` and tests to work with this new structure but am holding off on pushing this since it is outside of the scope of this PR. I am happy to include it if requested. 

The only memory store that I havent created an update for is the `WeaviateMemoryStore`, since I dont have it installed at this time.

Comment (dmytrostruk) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py]: As a best practice, I think datetimes in database should be stored in UTC and converted on client side to appropriate timezone. In other words, it's a client responsibility to handle datetime conversion.

Semantic Kernel, as SDK, shouldn't be aware of timezones at all. That means, if customer uses UTC to store datetimes in database, then SK puts a record with timestamp in UTC. If customer decided to use local timezone to store information - then SK puts a record with timestamp in local timezone. In other words, we should pass timezone in the same format as we received from SDK user and we should not perform any conversions.

As an example, in case there will be some default logic, when we put some default timestamp (for example `DateTime.UtcNow`) - the timezone should always be in UTC. Although even in this case I would probably set default timestamp value as `null`, but it depends on scenario.

@cschadewitz I see some datetime conversion in `postgres_memory_store.py`, so I think we need to remove it, since it's SDK consumers responsibility. Please let me know if that makes sense.

We will probably need to make some changes in C# kernel to make timestamp field as timezone-agnostic, but it's definitely a scope of separate PR.

File: python/tests/integration/connectors/memory/test_postgres.py
Comment (dmytrostruk) [inline python/tests/integration/connectors/memory/test_postgres.py:108]: nit: We can rename `get_postgres_config` to `connection_string` and pass it directly to `PostgresMemoryStore`. Same applies to all other tests.

**Review by dmytrostruk — COMMENTED** [review]
@cschadewitz Thank you for all the hard work! I left some comments, but some of them can be definitely resolved in scope of follow-up PR. I think the most important one which should be resolved before merging to main is regarding timezones, please find the details under my comment. Thanks again!

Thanks again!

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py]: That is fine with me if we are passing the responsibility off to the consumer of the SDK.

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:110]: Yeah this one is kind of tough for me since most of the memory store backends require a dimensionality at time of collection creation. I understand that the `MemoryStoreBase` doesn't have it as a parameter and as such the Kernel will not pass anything for it but I wonder if that should be the case. 

I wonder if there is a use case in which a consumer would interface with 2 embedding services (that return embeddings with different dimensionality). As the memory store stands now, they would need two instances of a memory store to support both dimensionalities, which I dont know if the Kernel even supports.

Food for thought.

Regardless, I can move the check into a single method but will maintain the parameter on the `create_collection_async` method for the time being

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:76]: Since Postgres does not store the timezone with a timestamp, I am going to remove all timezone related work and assume that the consumer is always passing in UTC time (or really whatever timezone they want) and are handling all necessary conversions

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:288]: fixed

File: python/tests/integration/connectors/memory/test_postgres.py
Comment (cschadewitz) [reply python/tests/integration/connectors/memory/test_postgres.py:108]: fixed

File: python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py
Comment (cschadewitz) [reply python/semantic_kernel/connectors/memory/postgres/postgres_memory_store.py:158]: Sounds fair, I will look at doing this in a subsequent pr

**Comment by awharrison-28** [comment]
Hi @cschadewitz - I'm setting up infra to be able to run the integration tests in CI. As soon as that's set up, we'll get this in asap.

**Review by awharrison-28 — APPROVED** [review]

**Comment by robertmujica** [comment]
Great work @cschadewitz , wondering if there is any pre-release i can use to test it ?

**Comment by cschadewitz** [comment]
> Great work @cschadewitz , wondering if there is any pre-release i can use to test it ?

Funny you should ask, I made a request for python to get nightly/pre-release builds during yesterdays community hour. Here is another community member that want pre-release pip packages @alexchaomander . 
Thanks!

**Pull request merged**
